Grammar and spelling fixes (#5229)
kubernetes/ingress-nginx: docs/enhancements/20190815-zone-aware-routing.md

## Motivation

> When users run their services across multiple availability zones they usually pay for egress traffic between zones. Providers such as GCP, Amazon EC2 charge money for that.

Author: https://github.com/kubernetes/ingress-nginx/pull/4455#discussion_r314538964

> When users run their services across multiple availability zones they usually pay for egress traffic between zones. Providers such as GCP, Amazon EC2 charge money for that.

Reviewer: ```suggestion
When users run their services across multiple availability zones they usually pay for egress traffic between zones. Providers such as GCP, and Amazon EC2 usually charge extra for this feature.
```

ingress-nginx when picking an endpoint to route request to does not consider whether the endpoint is in a different zone or the same one. That means it's at least equally likely

> that it will pick an endpoint from another zone and proxy the request to it. In this situation response from the endpoint to the ingress-nginx pod is considered as

Reviewer: ```suggestion
that it will pick an endpoint from another zone and proxy the request to it. In this situation response from the endpoint to the ingress-nginx pod is considered
```

> inter-zone traffic and costs money.

Reviewer: ```suggestion
inter-zone traffic and usually costs extra money.
```

## Proposal

The idea here is to have the controller part of ingress-nginx
(1) detect what zone its current pod is running in and
(2) detect the zone for every endpoint it knows about.
After that, it will post that data as part of endpoints to Lua land.

> Then Lua balancer when picking an endpoint will try to pick zone-local endpoint first and

Reviewer: ```suggestion
When picking an endpoint, the Lua balancer will try to pick zone-local endpoint first and
```

if there is no zone-local endpoint then it will fall back to current behaviour.

> This feature at least, in the beginning, should be optional since it is going to make it harder to reason about the load balancing and not everyone might want that.

Reviewer: ```suggestion
Initially, this feature should be optional since it is going to make it harder to reason about the load balancing and not everyone might want that.
```

**How does controller know what zone it runs in?**
We can have the pod spec pass the node name using downward API as an environment variable.

> Then on start controller can get node details from the API based on the node name.

Reviewer: ```suggestion
Upon startup, the controller can get node details from the API based on the node name.
```

Once the node details are obtained

> we can extract the zone from `failure-domain.beta.kubernetes.io/zone` annotation.

Reviewer: ```suggestion
we can extract the zone from the `failure-domain.beta.kubernetes.io/zone` annotation.
```

Then we can pass that value to Lua land through Nginx configuration
when loading `lua_ingress.lua` module in `init_by_lua` phase.

**How do we extract zones for endpoints?**
We can have the controller watch create and update events on nodes in the entire cluster and based on that keep the map of nodes to zones in the memory.
And when we generate endpoints list, we can access node name using `.subsets.addresses[i].nodeName`
and based on that fetch zone from the map in memory and store it as a field on the endpoint.
__This solution assumes `failure-domain.beta.kubernetes.io/zone`__ annotation does not change until the end of the node's life. Otherwise, we have to
watch update events as well on the nodes and that'll add even more overhead.

Alternatively, we can get the list of nodes only when there's no node in the memory for the given node name. This is probably a better solution
because then we would avoid watching for API changes on node resources. We can eagerly fetch all the nodes and build node name to zone mapping on start.

> And from thereon sync it during endpoint building in the main event loop IFF there's no existing entry for the node of an endpoint.

Reviewer: ```suggestion
From thereon,  it will sync during endpoint building in the main event loop if there's no existing entry for the node of an endpoint.
```

Author: I'm not a huge fan of `thereon`, is it really preferred over `there on`?

Reviewer: there on is probably better

Author: ```suggestion
From there on,  it will sync during endpoint building in the main event loop if there's no existing entry for the node of an endpoint.
```
